### ISSUE TITLE
apf will reset when panic happen

### ISSUE THREAD

**Issue by leileiwan** [kubernetes/kubernetes#111852]
### What happened?

when we use apf in our 5k production environment,  apiserver often freezes and cannot process any request because of the timeout, which results in many node being not ready.
we found apf will reset when http write panic, it took 10m(too long) process each pl,  many nodes become notready，then apiserver restart by probe.

```
cat log_import | grep 'Observed a panic: runtime error: invalid memory address or nil pointer dereference' -A 100

E0810 02:50:37.892721       1 runtime.go:76] Observed a panic: runtime error: invalid memory address or nil pointer dereference
goroutine 969509562 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1.1(0xc167486f60)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:108 +0x107
panic(0x3e43f20, 0x79b0880)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
compress/gzip.(*Writer).Write(0xc10c78ed10, 0xc227acfa80, 0x71, 0x71, 0x30, 0x41269c0, 0xc006019501)
        /usr/local/go/src/compress/gzip/gzip.go:168 +0x237
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.(*deferredResponseWriter).Write(0xc29477cbe0, 0xc227acfa80, 0x71, 0x71, 0xc227acfa80, 0x71, 0x71)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:182 +0x54e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject(0x46c9bcb, 0x10, 0x7f540b9edd50, 0xc2606f2c80, 0x52f7f00, 0xc00335a508, 0xc2dec2f200, 0xc8, 0x529be40, 0xc204895c70)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:117 +0x389
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x52fb200, 0xc0028baae0, 0x52f7e80, 0xc0b76ae340, 0x0, 0x0, 0x46ac6ea, 0x2, 0x52f7f00, 0xc00335a508, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:251 +0x555
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.transformResponseObject(0x5305e40, 0xc250341c80, 0xc0b76ae340, 0xc00601a168, 0xc2dec2f200, 0x52f7f00, 0xc00335a508, 0xc8, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/response.go:137 +0x353
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1(0x52f7f00, 0xc00335a508, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go:276 +0xdfc
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints.restfulListResource.func1(0xc250341bc0, 0xc204895c00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/installer.go:1092 +0x8f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc250341bc0, 0xc204895c00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:356 +0x254
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc000e721b0, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:288 +0xa4f
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).Dispatch(...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:199
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x46c35d7, 0xe, 0xc000e721b0, 0xc0007da620, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:146 +0x4d3
k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).ServeHTTP(0xc04a03b340, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:121 +0x161
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc0866decc0, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x38a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc0ecd19d50, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x84
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x46c6a64, 0xf, 0xc0ecd54bd0, 0xc0ecd19d50, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:154 +0x6b1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:64 +0x512
net/http.HandlerFunc.ServeHTTP(0xc0ecd03440, 0x52f8080, 0xc00335a4a8, 0xc2dec2f200)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func2.4()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:139 +0x1d3
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Handle.func2()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:177 +0x22e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol.immediateRequest.Finish(...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:764
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Handle(0xc00026a000, 0x5305e40, 0xc250341a10, 0xc16b4ca4d0, 0x5306880, 0xc1516e0380, 0xc1a7ef29b0, 0xc1a7ef29c0, 0xc187f2a780)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:167 +0x76d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func2(0x52f8080, 0xc00335a4a8, 0xc2dec2f100)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:142 +0x508
net/http.HandlerFunc.ServeHTTP(0xc0ecd68480, 0x52f8080, 0xc00335a4a8, 0xc2dec2f100)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithLongRunningMonitor.func1(0x52f8080, 0xc00335a4a8, 0xc2dec2f100)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/request_monitor.go:46 +0x195
net/http.HandlerFunc.ServeHTTP(0xc0ecd52ea0, 0x52f8080, 0xc00335a4a8, 0xc2dec2f100)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithTkexCurrentLimit.func1(0x52f8080, 0xc00335a4a8, 0xc2dec2f100)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/request_monitor.go:113 +0x790
net/http.HandlerFunc.ServeHTTP(0xc0ecd684b0, 0x7f53fd166048, 0xc00335a498, 0xc2dec2f100)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x7f53fd166048, 0xc00335a498, 0xc2dec2f100)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:50 +0x1fe6
net/http.HandlerFunc.ServeHTTP(0xc0ecd03480, 0x7f53fd166048, 0xc00335a498, 0xc2dec2f100)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f53fd166048, 0xc00335a498, 0xc2dec2ef00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:70 +0x5ce
net/http.HandlerFunc.ServeHTTP(0xc0eccfae10, 0x7f53fd166048, 0xc00335a498, 0xc2dec2ef00)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc167486f60, 0xc0ecd52f60, 0x5306940, 0xc00335a498, 0xc2dec2ef00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:113 +0xd0
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:99 +0x1cb

goroutine 969509561 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3b68780, 0xc1f18cacd0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc1bfbbfc90, 0x1, 0x1)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x3b68780, 0xc1f18cacd0)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc0ecd52f60, 0x52f80c0, 0xc204895b20, 0xc2dec2ef00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:119 +0x486
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithWaitGroup.func1(0x52f80c0, 0xc204895b20, 0xc2dec2ee00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/waitgroup.go:59 +0x121
net/http.HandlerFunc.ServeHTTP(0xc0ecd684e0, 0x52f80c0, 0xc204895b20, 0xc2dec2ee00)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x52f80c0, 0xc204895b20, 0xc2dec2ed00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x274
net/http.HandlerFunc.ServeHTTP(0xc0ecd68510, 0x52f80c0, 0xc204895b20, 0xc2dec2ed00)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithCacheControl.func1(0x52f80c0, 0xc204895b20, 0xc2dec2ed00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/cachecontrol.go:31 +0xa8
net/http.HandlerFunc.ServeHTTP(0xc0ecd52f80, 0x52f80c0, 0xc204895b20, 0xc2dec2ed00)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.WithLogging.func1(0x52ea880, 0xc00335a488, 0xc2dec2ec00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:89 +0x2ca
net/http.HandlerFunc.ServeHTTP(0xc0ecd52fa0, 0x52ea880, 0xc00335a488, 0xc2dec2ec00)
        /usr/local/go/src/net/http/server.go:2007 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1(0x52ea880, 0xc00335a488, 0xc2dec2ec00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:51 +0x13e
net/http.HandlerFunc.ServeHTTP(0xc0ecd52fc0, 0x52ea880, 0xc00335a488, 0xc2dec2ec00)
```

```
I0810 02:50:37.892553       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 02:50:37.892652       1 apf_controller.go:359] Controller syncOne at 2022-08-10T02:50:37.892
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:764
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:764
I0810 02:50:37.894155       1 apf_controller.go:660] Retaining queues for priority level "catch-all": config={"type":"Limited","limited":{"assuredConcurrencyShares":5,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=112, quiescing=false, numPending=0 (shares=5, shareSum=225)
I0810 02:53:01.492398       1 apf_controller.go:647] Using exempt priority level "exempt": quiescing=false
I0810 02:53:01.492416       1 apf_controller.go:660] Retaining queues for priority level "global-default": config={"type":"Limited","limited":{"assuredConcurrencyShares":80,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=1778, quiescing=false, numPending=0 (shares=80, shareSum=225)
I0810 02:54:14.632115       1 apf_controller.go:660] Retaining queues for priority level "lbcf": config={"type":"Limited","limited":{"assuredConcurrencyShares":20,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=445, quiescing=false, numPending=0 (shares=20, shareSum=225)
I0810 02:56:32.764250       1 apf_controller.go:660] Retaining queues for priority level "leader-election": config={"type":"Limited","limited":{"assuredConcurrencyShares":10,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=223, quiescing=false, numPending=0 (shares=10, shareSum=225)
I0810 02:57:46.991882       1 apf_controller.go:660] Retaining queues for priority level "log-metrics": config={"type":"Limited","limited":{"assuredConcurrencyShares":10,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=223, quiescing=false, numPending=0 (shares=10, shareSum=225)
I0810 03:00:13.539033       1 apf_controller.go:660] Retaining queues for priority level "system": config={"type":"Limited","limited":{"assuredConcurrencyShares":60,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=1334, quiescing=false, numPending=0 (shares=60, shareSum=225)
I0810 03:01:23.959826       1 apf_controller.go:660] Retaining queues for priority level "workload-high": config={"type":"Limited","limited":{"assuredConcurrencyShares":40,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=889, quiescing=false, numPending=0 (shares=40, shareSum=225)
I0810 03:02:49.184108       1 apf_controller.go:508] Switched to new API Priority and Fairness configuration
I0810 03:02:49.191681       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 03:02:49.192208       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 03:02:49.198052       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 03:02:49.211140       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 03:02:49.211214       1 apf_controller.go:844] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0810 03:02:49.216658       1 apf_controller.go:359] Controller syncOne at 2022-08-10T03:02:49.216
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:764
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:764
I0810 03:02:53.365206       1 apf_controller.go:660] Retaining queues for priority level "catch-all": config={"type":"Limited","limited":{"assuredConcurrencyShares":5,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=112, quiescing=false, numPending=0 (shares=5, shareSum=225)
I0810 03:02:54.388745       1 apf_controller.go:647] Using exempt priority level "exempt": quiescing=false
I0810 03:02:54.388763       1 apf_controller.go:660] Retaining queues for priority level "global-default": config={"type":"Limited","limited":{"assuredConcurrencyShares":80,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=1778, quiescing=false, numPending=0 (shares=80, shareSum=225)
I0810 03:02:54.825738       1 apf_controller.go:660] Retaining queues for priority level "lbcf": config={"type":"Limited","limited":{"assuredConcurrencyShares":20,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=445, quiescing=false, numPending=0 (shares=20, shareSum=225)
I0810 03:02:55.497572       1 apf_controller.go:660] Retaining queues for priority level "leader-election": config={"type":"Limited","limited":{"assuredConcurrencyShares":10,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=223, quiescing=false, numPending=0 (shares=10, shareSum=225)
I0810 03:02:55.782596       1 apf_controller.go:660] Retaining queues for priority level "log-metrics": config={"type":"Limited","limited":{"assuredConcurrencyShares":10,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=223, quiescing=false, numPending=0 (shares=10, shareSum=225)
I0810 03:02:56.180468       1 apf_controller.go:660] Retaining queues for priority level "system": config={"type":"Limited","limited":{"assuredConcurrencyShares":60,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=1334, quiescing=false, numPending=0 (shares=60, shareSum=225)
I0810 03:02:56.297450       1 apf_controller.go:660] Retaining queues for priority level "workload-high": config={"type":"Limited","limited":{"assuredConcurrencyShares":40,"limitResponse":{"type":"Reject"}}}, concurrencyLimit=889, quiescing=false, numPending=0 (shares=40, shareSum=225)
I0810 03:02:56.378736       1 apf_controller.go:508] Switched to new API Priority and Fairness configuration
```

### What did you expect to happen?

apf should not reset when panic happens

### How can we reproduce it (as minimally and precisely as possible)?

```
```

### Anything else we need to know?

```
```

### Kubernetes version

```
```

### Cloud provider

```
```

### OS version

```
```

### Install tools

```
```

### Container runtime (CRI) and version (if applicable)

```
```

### Related plugins (CNI, CSI, ...) and versions (if applicable)

```
```

**Comment by MikeSpreitzer** [comment]
Which commit of APF was involved in the trace shown in the opening comment?

I see that the log messages about digesting config are dated 9 days later than the log message about panic.  Why is it asserted that these two are related?

The fact that it took about 12 minutes to digest the config (starting at 0810 02:50:37) is quite surprising.  This should only be a matter of local computation, I do not see why it should take 12 minutes.  There is a big mutex held during that time, so this is a real problem.

**Comment by leileiwan** [comment]
> Which commit of APF was involved in the trace shown in the opening comment?
> 
> I see that the log messages about digesting config are dated 9 days later than the log message about panic. Why is it asserted that these two are related?
> 
> The fact that it took about 12 minutes to digest the config (starting at 0810 02:50:37) is quite surprising. This should only be a matter of local computation, I do not see why it should take 12 minutes. There is a big mutex held during that time, so this is a real problem.

I've updated the first panic log, the failure frequency is very high in our production environment.
Now we solved the problem by fixing apf reset bug when panic happens,  but mutex also needs to be optimized.

**Comment by MikeSpreitzer** [comment]
It would help a lot to know the version of the apiserver involved here.

**Comment by leileiwan** [comment]
we rebuild apiserver with golang
```
go version go1.16.13 linux/386
```

```
 kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4-tke.20.54+69eb45f2ef7d63", GitCommit:"69eb45f2ef7d63121c92ccb9e7b2af40c6ef9372", GitTreeState:"dirty", BuildDate:"2022-08-10T06:34:11Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

```
Client: Docker Engine - Community
 Version:           19.03.9
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        9d988398e7
 Built:             Fri May 15 00:22:47 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.9
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       9d988398e7
  Built:            Fri May 15 00:28:17 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by MikeSpreitzer** [comment]
This is using Kubernetes release 1.18.4?  That's really old!

**Comment by aojea** [comment]
> This is using Kubernetes release 1.18.4? That's really old!

indeed, @leileiwan bear in mind that only supported releases https://kubernetes.io/releases/ can receive bugfixes

**Comment by MikeSpreitzer** [comment]
FYI, there is a real bug here that persisted until now; the fix is #111850 .  But @leileiwan should not expect the fix to be back-ported to release 1.18.
